worlds/mmx/Client.py

```diff
@@ -0,0 +1,1112 @@
+        can_move = await snes_read(ctx, MMX_CAN_MOVE, 0x7)
+        pause_state = await snes_read(ctx, MMX_PAUSE_STATE, 0x1)
+        if menu_state[0] != 0x04 or \
+            gameplay_state[0] != 0x04 or \
+            pause_state[0] != 0x00 or \
+            can_move != b'\x00\x00\x00\x00\x00\x00\x00' or \
+            receiving_item[0] != 0x00:
+            return
+        
+        snes_buffered_write(ctx, MMX_CURRENT_HP, bytes([0x80]))
+        snes_buffered_write(ctx, WRAM_START + 0x00BAA, bytes([0x0C]))
+        snes_buffered_write(ctx, WRAM_START + 0x00C12, bytes([0x0C]))
+        snes_buffered_write(ctx, WRAM_START + 0x00BAB, bytes([0x00]))
+        snes_buffered_write(ctx, WRAM_START + 0x00BD7, bytes([0x08]))
+
+        await snes_flush_writes(ctx)
+
+        ctx.death_state = DeathState.dead
+        ctx.last_death_link = time.time()
+
+
+    async def validate_rom(self, ctx):
+        from SNIClient import snes_read
+
+        energy_link = await snes_read(ctx, MMX_ENERGY_LINK_ENABLED, 0x1)
+        rom_name = await snes_read(ctx, MMX_ROMHASH_START, ROMHASH_SIZE)
+        if rom_name is None or rom_name == bytes([0] * ROMHASH_SIZE) or rom_name[:4] != b"MMX1":
+            if "resync" in ctx.command_processor.commands:
+                ctx.command_processor.commands.pop("resync")
+            if "trade" in ctx.command_processor.commands:
+                ctx.command_processor.commands.pop("trade")
+            if "heal" in ctx.command_processor.commands:
+                ctx.command_processor.commands.pop("heal")
+            if "refill" in ctx.command_processor.commands:
+                ctx.command_processor.commands.pop("refill")
+            return False
+        
+        ctx.game = self.game
+        ctx.items_handling = 0b111
+        ctx.receive_option = 0
+        ctx.send_option = 0
+        ctx.allow_collect = True
+        if energy_link[0]:
+            if "refill" not in ctx.command_processor.commands:
+                ctx.command_processor.commands["heal"] = cmd_heal
+            if "refill" not in ctx.command_processor.commands:
+                ctx.command_processor.commands["refill"] = cmd_refill
+        if "resync" not in ctx.command_processor.commands:
+            ctx.command_processor.commands["resync"] = cmd_resync
+        if "trade" not in ctx.command_processor.commands:
+            ctx.command_processor.commands["trade"] = cmd_trade
+
+        death_link = await snes_read(ctx, MMX_DEATH_LINK_ACTIVE, 1)
+        if death_link[0]:
+            await ctx.update_death_link(bool(death_link[0] & 0b1))
+
+        ctx.rom = rom_name
+
+        return True
+     
+    
+    async def game_watcher(self, ctx):
+        from SNIClient import snes_buffered_write, snes_flush_writes, snes_read
+
+        game_state = await snes_read(ctx, MMX_GAME_STATE, 0x1)
+        menu_state = await snes_read(ctx, MMX_MENU_STATE, 0x1)
+        gameplay_state = await snes_read(ctx, MMX_GAMEPLAY_STATE, 0x1)
```

Comment:
In general, all the code in this client is treating `snes_read` as if it's as fast and as cheap and as reliable as reading from your own ram.
The `await` should be a clue that that isn't the case.
I think it's better to treat `snes_read` like asking someone to go to the other side of town to get something for you.

In this example, you're asking someone to cross town to get `MMX_GAME_STATE`,
and then after they return, you immediate ask them to go back, to the same place they just came from, to get another small thing `MMX_MENU_STATE` that was sitting right next to the first thing.
The person would probably say "Why didn't you tell me to get that in the first trip?"
And then you ask them to go get a 3rd single byte that's right next to the first two.

I would recommend reworking most of the `snes_read` to just get a few bigger blocks of memory that have everything you need all at once.

I was surprised by how much faster it could be getting a block of memory at once instead of individual bytes one at a time.
https://discord.com/channels/731205301247803413/1022545979146252288/1239239492385247252
<details>(This link is not meant to indicate that FF6 is a good example to follow, because there are still a lot of patterns that aren't so good. It's just to see players noticing and talking about the amount of time things take.)</details>

Reply:
It's even better to get some memory that you don't need to make fewer trips.
example:
If you need 101, 102, 117, and 118
it's probably better to just get all of 101-118
even though you don't need 103-116

Reply:
Another reason to do this rework to read memory in fewer bigger chunks:

The more times someone has to travel across town, the more likely they are to get in an accident and not return.

```bash
$ python -m pip install --upgrade mypy
$ python -m mypy --check-untyped-defs --follow-imports=silent --allow-untyped-def --disable-error-code=assignment --disable-error-code=operator --disable-error-code=comparison-overlap --disable-error-code=attr-defined worlds/mmx/Client.py
```

This reports 104 errors and almost all of these are real bugs that will happen regularly.
I have seen from someone else an indication that MMX has one of the most unstable clients of all SNI clients.
From this, I can see why.
<details>Run the same command on Super Metroid's `Client.py`, and you can see that it has only 1 error report.</details>

I recommend that you not try to handle these mypy error reports until after the rework that I suggested, because afterwards there will be fewer error reports to handle.

Reply:
Ahhh, that makes so much sense with some reports I've received before. I do assume those show up less frequently in powerful machines, as I've not seen my client(s) being unstable on my end at all.

I'm gonna rework X1's client (and the other two as well, they're pretty much the same code), I believe I can group several of those as most of the RAM reads regarding collected locations and upgrades are _almost_ in the same block of RAM.